File: 힙/프로그래머스lv2-더 맵게.py
import heapq
def solution(scoville, K):
    heap = scoville[:]
    heapq.heapify(heap)
    
    count = 0
    while heap[0] < K:
        if len(heap) < 2:
            return -1
        first, second = heapq.heappop(heap), heapq.heappop(heap)
        new = first + 2 * second
        count += 1
        heapq.heappush(heap, new)
        
    return count

"""
정확성  테스트
테스트 1 〉	통과 (0.01ms, 10.2MB)
테스트 2 〉	통과 (0.01ms, 10.2MB)
테스트 3 〉	통과 (0.01ms, 10.2MB)
테스트 4 〉	통과 (0.01ms, 10.2MB)
테스트 5 〉	통과 (0.01ms, 10.2MB)
테스트 6 〉	통과 (0.79ms, 10.3MB)
테스트 7 〉	통과 (0.60ms, 10.2MB)
테스트 8 〉	통과 (0.05ms, 10.3MB)
테스트 9 〉	통과 (0.07ms, 10.2MB)
테스트 10 〉	통과 (0.36ms, 10.2MB)
테스트 11 〉	통과 (0.20ms, 10.2MB)
테스트 12 〉	통과 (1.32ms, 10.2MB)
테스트 13 〉	통과 (0.68ms, 10.2MB)
테스트 14 〉	통과 (0.02ms, 10.2MB)
테스트 15 〉	통과 (0.59ms, 10.4MB)
테스트 16 〉	통과 (0.01ms, 10.2MB)
효율성  테스트
테스트 1 〉	통과 (162.28ms, 18.2MB)
테스트 2 〉	통과 (376.33ms, 25.6MB)
테스트 3 〉	통과 (1795.25ms, 62.1MB)
테스트 4 〉	통과 (138.14ms, 16.7MB)
테스트 5 〉	통과 (1944.68ms, 64.8MB)
"""

File: 힙/프로그래머스lv2-더 맵게.js
// 자바스크립트를 지원하지 않는 문제라서 직접 테스트케이스 넣기..
const TEST_CASE = [[[1, 2, 3, 9, 10, 12], 7, 2]]

class BinaryHeap {
  constructor() {
    this.items = [null]
  }

  get length() {
    return this.items.length - 1
  }

  get head() {
    if (this.items.length < 2) return undefined
    return this.items[1]
  }

  // insert heapify
  _percolateUp() {
    let i = this.length
    let parent = Math.floor(i / 2)
    while (parent > 0) {
      if (this.items[i] < this.items[parent]) {
        let temp = this.items[parent]
        this.items[parent] = this.items[i]
        this.items[i] = temp
      }
      i = parent
      parent = Math.floor(i / 2)
    }
  }

  insert(k) {
    this.items.push(k)
    this._percolateUp()
  }

  // pop heapify
  _percolateDown(idx) {
    let left = idx * 2
    let right = idx * 2 + 1
    let smallest = idx

    if (left <= this.length && this.items[left] < this.items[smallest]) {
      smallest = left
    }
    if (right < this.length && this.items[right] < this.items[smallest]) {
      smallest = right
    }
    if (smallest !== idx) {
      let temp = this.items[idx]
      this.items[idx] = this.items[smallest]
      this.items[smallest] = temp
      this._percolateDown(smallest)
    }
  }

  extract() {
    let extracted = this.items[1]
    this.items[1] = this.items[this.length]
    this.items.pop()
    this._percolateDown(1)
    return extracted
  }
}

function solution(scoville, K, _return) {
  const heap = new BinaryHeap()
  for (const x of scoville) {
    heap.insert(x)
  }
  let count = 0

  while (heap.head < K) {
    if (heap.length < 2) return -1
    const first = heap.extract()
    const second = heap.extract()
    const _new = first + 2 * second
    count++
    heap.insert(_new)
  }
  return count
}

for (const [scoville, K, _return] of TEST_CASE) {
  const result = solution(scoville, K, _return)
  console.log(result === _return)
}

// true
